File: clients/python/examples/crowd.py
"""Crowdsourced pairwise votes with unreliable annotators (Crowd-BT).

Three annotators judge the same items with very different reliability:
alice is diligent, bob is a coin-flipping spammer, mallory votes against
the truth. Crowd-BT weights each annotator's votes by an estimated
reliability, so mallory's inverted votes *add* signal instead of poisoning
the fit. Contrast with plain Bradley-Terry that ignores who voted.

Run after `maturin develop`:  python examples/crowd.py
"""

import random

import propagon

TRUE_ORDER = ["sonnet", "haiku", "limerick", "ballad", "ode", "epigram"]


def votes():
    """Deterministic synthetic votes: (annotator, winner, loser)."""
    rng = random.Random(2024)
    annotators = [("alice", 0.92), ("bob", 0.50), ("mallory", 0.12)]
    for _ in range(200):
        for name, fidelity in annotators:
            i, j = sorted(rng.sample(range(len(TRUE_ORDER)), 2))
            better, worse = TRUE_ORDER[i], TRUE_ORDER[j]  # i<j => better truth
            if rng.random() < fidelity:
                yield name, better, worse
            else:
                yield name, worse, better


def main() -> None:
    crowd = propagon.AnnotatedPairsDataset()
    naive = propagon.PairwiseDataset()
    for annotator, winner, loser in votes():
        crowd.push(annotator, winner, loser)
        naive.push(winner, loser)  # same votes, annotator column dropped

    print(f"True order: {' > '.join(TRUE_ORDER)}\n")

    print("Crowd-BT (reliability-aware):")
    for name, score in propagon.CrowdBt().fit(crowd).sorted_scores():
        print(f"  {name:10s} {score:7.4f}")

    print("\nPlain Bradley-Terry (ignores who voted):")
    for name, score in propagon.BradleyTerryMM().fit(naive).sorted_scores():
        print(f"  {name:10s} {score:7.4f}")


if __name__ == "__main__":
    main()

File: clients/wasm/examples/crowd.ts
// Crowdsourced votes with unreliable annotators: Crowd-BT vs plain Bradley-Terry.
//   node examples/crowd.ts
import { datasets, annotated, pairwise } from "../index.js";

const TRUE_ORDER = ["sonnet", "haiku", "limerick", "ballad", "ode", "epigram"];

// Deterministic synthetic votes from a diligent, a spammer, and an adversary.
function mulberry32(a: number): () => number {
  return () => {
    a = (a + 0x6d2b79f5) | 0;
    let t = Math.imul(a ^ (a >>> 15), 1 | a);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

function* votes(): Generator<[string, string, string]> {
  const rand = mulberry32(2024);
  // A diligent annotator and an adversary who votes against the truth. Plain
  // Bradley-Terry sees the two cancel out; Crowd-BT detects that mallory is
  // inverted and turns her votes back into signal.
  const annotators: [string, number][] = [["alice", 0.9], ["mallory", 0.1]];
  for (let n = 0; n < 400; n++) {
    for (const [name, fidelity] of annotators) {
      let i = Math.floor(rand() * TRUE_ORDER.length);
      let j = Math.floor(rand() * TRUE_ORDER.length);
      if (i === j) continue;
      if (i > j) [i, j] = [j, i];
      const [better, worse] = [TRUE_ORDER[i], TRUE_ORDER[j]];
      yield rand() < fidelity ? [name, better, worse] : [name, worse, better];
    }
  }
}

const crowd = new datasets.AnnotatedPairsDataset();
const naive = new datasets.PairwiseDataset();
for (const [annotator, winner, loser] of votes()) {
  crowd.push(annotator, winner, loser, 1);
  naive.push(winner, loser, 1);
}

console.log(`True order: ${TRUE_ORDER.join(" > ")}\n`);
console.log("Crowd-BT (reliability-aware):");
for (const [name, score] of annotated.fitCrowdBt({}, crowd).sortedScores()) {
  console.log(`  ${name.padEnd(10)} ${score.toFixed(4)}`);
}
console.log("\nPlain Bradley-Terry (ignores who voted):");
for (const [name, score] of pairwise.fitBradleyTerryMm({}, naive).sortedScores()) {
  console.log(`  ${name.padEnd(10)} ${score.toFixed(4)}`);
}
